### PR TITLE
feat: S12.3 buffer overflow protection at FormatCharacter

### LINKED ISSUE #87: Story: TDD-driven buffer overflow protection at FormatCharacter
**Parent epic:** #12 (E10: Static Analysis and MISRA)
**Depends on:** #86 (Extract Formatter)

## Motivation

After the Formatter extraction (#86), all writes funnel through FormatCharacter.
This story adds buffer overflow protection at that single chokepoint via TDD.
When the buffer is full, FormatCharacter silently discards further writes —
the message is truncated but no memory corruption occurs.

### FormatCharacter with overflow protection

```c
static inline size_t FormatCharacter(struct SolidSyslogFormatter* formatter, char value)
{
    if (formatter->position >= formatter->size - 1)
    {
        return 0;
    }
    formatter->buffer[formatter->position] = value;
    formatter->position++;
    formatter->buffer[formatter->position] = '\0';
    return 1;
}
```

Returns 0 when buffer is full. Callers already accumulate return values in `len`,
so truncation propagates naturally — subsequent format calls write nothing.

### TDD sequence

1. FormatCharacter writes character at position (existing behaviour, new test)
2. FormatCharacter returns 1 on success
3. FormatCharacter at buffer boundary returns 0
4. FormatCharacter at buffer boundary does not write
5. FormatCharacter always null-terminates
6. Full message truncation — Log with tiny buffer produces truncated but valid prefix
7. BoundedString respects remaining capacity

### Scope

- `Source/SolidSyslogFormatter.c` — FormatCharacter overflow check
- `Tests/SolidSyslogFormatterTest.cpp` — new test file for FormatCharacter
- Potentially FormatBoundedString overflow check
- No changes to public API

### MISRA C:2012 alignment

- Addresses Rule 1.3 (no undefined behaviour)
- Addresses Rule 21.17/21.18 (string handling bounds)
- FormatCharacter becomes the single verified write primitive

### PR DESCRIPTION
**Pull request by DavidCozens**
## Purpose

Story S12.3 (closes #87). TDD-driven buffer overflow protection at the Formatter's
single write point (WriteChar). Every buffer write is now bounded — writes past capacity
are silently stopped and the null terminator is always maintained.

## Change Description

**Overflow protection** — `HasCapacity` guard in `WriteChar` checks both `size > 0` and
`position < size - 1` before every write. This is the single enforcement point for all
Formatter output.

**Write functions return void** — `Character`, `BoundedString`, `Uint32`, `PaddedUint32`
no longer return sizes. No caller used the return values; callers that need to know if
content was written use the `Length` before/after pattern.

**`Remaining` removed from public API** — with bounds checking in WriteChar, callers no
longer need to pre-check capacity. `FormatMsg` uses `SOLIDSYSLOG_MAX_MESSAGE_SIZE` as the
maxLength for the message body, relying on WriteChar to stop at the buffer boundary.

**FileStore pathPrefix reservation** — `FormatFilename` now caps the pathPrefix to
`MAX_PREFIX_LENGTH` (= `MAX_PATH_SIZE - sequence digits - extension - null`), preventing
long prefixes from collapsing different sequence numbers to the same truncated filename.

**FormatMsgId** — converted from return-value check to Length before/after pattern,
consistent with FormatStructuredData and FormatStringField.

## Test Evidence

11 new boundary tests covering all write functions:
- `ZeroSizeBufferDoesNotCrash` — Character, BoundedString, Uint32 on zero-size formatter
- `CharacterFillsExactlyOneByteBuffer` — 1-byte buffer = null terminator only
- `CharacterStopsWhenFull` — writes beyond capacity silently dropped
- `BoundedStringStopsAtBufferCapacity` — truncates at buffer boundary
- `BoundedStringWritesNothingWhenFull` — second write to full buffer is no-op
- `Uint32TruncatedWhenBufferTooSmall` — partial digit output
- `Uint32FitsExactly` — value fills buffer exactly
- `PaddedUint32TruncatedWhenBufferTooSmall` — padding truncated at boundary
- `PaddedUint32FitsExactly` — padded value fills buffer exactly

467 tests pass across debug, clang, sanitize. Coverage 99.9%.
clang-tidy, cppcheck, clang-format clean. BDD 31/31 scenarios pass.

## Areas Affected

- `Interface/SolidSyslogFormatter.h` — write functions return void, Remaining removed
- `Source/SolidSyslogFormatter.c` — void returns, Remaining removed
- `Source/SolidSyslog.c` — FormatMsgId uses Length pattern, FormatMsg drops Remaining
- `Source/SolidSyslogFileStore.c` — pathPrefix capped to reserve suffix space
- `Tests/SolidSyslogFormatterTest.cpp` — 11 new boundary tests, return value checks removed
- `CLAUDE.md` — Remaining removed from header table

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Refactor**
  * Simplified formatter API: output operations no longer return write counts.
  * Removed the buffer-capacity query.
  * Renamed raw-data accessor to a clearer string-view accessor.
  * Added new fixed-width digit formatters (2-, 4-, 6-digit) and updated codepaths to use them.

* **Tests**
  * Added comprehensive tests for buffer edge cases, truncation, and zero-size behavior.
  * Updated tests to validate output via the new string-view accessor and length checks.

### CODE REVIEW

File: Tests/SolidSyslogFormatterTest.cpp [hunk 1]
```diff
@@ -15,176 +30,243 @@ TEST_GROUP(SolidSyslogFormatter)
...
     {
         // cppcheck-suppress unreadVariable -- formatter is used across TEST() bodies via CppUTest macro
-        formatter = SolidSyslogFormatter_Create(storage, TEST_BUFFER_SIZE);
+        CREATE_FORMATTER(TEST_BUFFER_SIZE);
     }
+
+    void FormatCharacter(char value) { SolidSyslogFormatter_Character(formatter, value); }
+    void FormatBoundedString(const char* source, size_t maxLength) { SolidSyslogFormatter_BoundedString(formatter, source, maxLength); }
+    void FormatUint32(uint32_t value) { SolidSyslogFormatter_Uint32(formatter, value); }
+    void FormatPaddedUint32(uint32_t value, size_t width) { SolidSyslogFormatter_PaddedUint32(formatter, value, width); }
```
[the remark this answers is not in the document]
Comment (DavidCozens) [reply Tests/SolidSyslogFormatterTest.cpp]: Fixed in 2fe02e0. Renamed all test helpers to camelCase: formatCharacter, formatBoundedString, formatUint32, formatTwoDigit, formatFourDigit, formatSixDigit.

File: Tests/SolidSyslogFormatterTest.cpp [hunk 2]
```diff
@@ -15,176 +30,243 @@ TEST_GROUP(SolidSyslogFormatter)
...
+}
+
+TEST(SolidSyslogFormatter, BoundedStringFillsExactCapacity)
+{
+    CREATE_FORMATTER(4);
+
+    FormatBoundedString("abcdef", 6);
+
+    CHECK_FORMATTED("abc");
+}
```
[the remark this answers is not in the document]
Comment (DavidCozens) [reply Tests/SolidSyslogFormatterTest.cpp:267]: Fixed in 2fe02e0. Changed to `formatBoundedString("abc", 3)` — a 3-char string that exactly fills the 3-char usable capacity of a 4-byte buffer.
